File: Coins2/app/src/main/java/com/lq/coins/MainActivity.java
package com.lq.coins;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.os.Bundle;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.lq.coins.databinding.ActivityMainBinding;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity {
    ActivityMainBinding binding;
    String BASE_URL = "https://raw.githubusercontent.com/";
    ArrayList<CoinModel> coins;
    CompositeDisposable disposable;
    ApiService service;
    Retrofit retrofit;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        disposable = new CompositeDisposable();

        Gson gson =new GsonBuilder()
                .setLenient()
                .create();

        retrofit =new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

         service = retrofit.create(ApiService.class);


        fetchData();


    }


    public void fetchData()
    {
        try{
            disposable.add( service.getAll()
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(this::handelResponse)
            );

        }
        catch (Exception e)
        {
            System.out.println(e.getLocalizedMessage());
        }
    }

    public void handelResponse(List<CoinModel> items)
    {
        coins = new ArrayList<CoinModel>(items);
        RecyclerAdapter adapter = new RecyclerAdapter(coins);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));
        binding.recyclerView.setAdapter(adapter);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        disposable.clear();
    }
}

File: CryptoAppWithKotlin/app/src/main/java/com/lq/cryptoappwithkotlin/view/MainActivity.kt
package com.lq.cryptoappwithkotlin.view

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import retrofit2.Converter.Factory
import androidx.recyclerview.widget.LinearLayoutManager
import com.lq.cryptoappwithkotlin.databinding.ActivityMainBinding
import com.lq.cryptoappwithkotlin.modal.CryptoModal
import com.lq.cryptoappwithkotlin.service.ApiService
import com.lq.cryptoappwithkotlin.utulitys.RecyclerAdapter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding;
    var items: ArrayList<CryptoModal>? = null;
    private lateinit var job: Job;
    lateinit var adapter: RecyclerAdapter;
    private val BASE_URL: String = "https://raw.githubusercontent.com/"
    private val colors = arrayListOf<String>(
        "#a6f5c5",
        "#fb7105",
        "#727cf5",
        "#94f667",
        "#d0cece",
        "#0a225b",
        "#d96ad5",
        "#dcd1b4"
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root);
        fetchData()
    }


    fun fetchData() {

        val retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build();
         val apiService = retrofit.create(ApiService::class.java);

         job = CoroutineScope(Dispatchers.IO).launch {
                apiService.getData().also {
                           withContext(Dispatchers.Main){
                               adapter = RecyclerAdapter(ArrayList<CryptoModal>(it),colors);
                               binding.recyclerView.layoutManager = LinearLayoutManager(this@MainActivity);
                               binding.recyclerView.adapter = adapter;
                           }
                }
         }



    }


    override fun onDestroy() {
        super.onDestroy()
        job.cancel()
    }

}